napari/layers/_data_protocols.py

```diff
@@ -0,0 +1,129 @@
+"""This module holds Protocols that layer.data objects are expected to provide.
+"""
+from __future__ import annotations
+
+from types import GeneratorType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+import numpy as np
+from typing_extensions import Protocol, runtime_checkable
+
+from .utils.layer_utils import compute_multiscale_level_and_corners
+
+_T = TypeVar('_T')
+Shape = Tuple[int, ...]
+ListOrTuple = Union[List[_T], Tuple[_T, ...], np.ndarray]
+_OBJ_NAMES = set(dir(Protocol))
+_OBJ_NAMES.update({'__annotations__', '__dict__', '__weakref__'})
+
+if TYPE_CHECKING:
+    from ..types import DTypeLike
+
+
+def _raise_protocol_error(obj: Any, protocol: type):
+    """Raise a more helpful error when required protocol members are missing."""
+    needed = set(dir(protocol)).union(protocol.__annotations__) - _OBJ_NAMES
+    missing = needed - set(dir(obj))
+    message = (
+        f"Object of type {type(obj).__name__!r} does not implement "
+        f"{protocol.__name__!r} Protocol.\n"
+        f"Missing methods: {missing!r}"
+    )
+    raise TypeError(message)
+
+
+@runtime_checkable
+class LayerDataProtocol(Protocol):
+    """A Protocol that all layer.data needs to support.
+
+    WIP: Shapes.data may be an execption.
+    """
+
+    shape: Shape
+    dtype: DTypeLike
+
+    def __getitem__(self, item) -> LayerDataProtocol:
+        ...
+
+
+def assert_protocol(obj: Any, protocol: Protocol = LayerDataProtocol):
+    """Assert `obj` is an instance of `protocol` or raise helpful error."""
+    if not isinstance(obj, protocol):
+        _raise_protocol_error(obj, protocol)
+
+
+class MultiScaleData(Sequence[LayerDataProtocol], LayerDataProtocol):
+    """Wrapper for multiscale data, to provide consistent API."""
+
+    def __init__(self, data, max_size: Optional[Sequence[int]] = None) -> None:
+        if isinstance(data, GeneratorType):
+            data = list(data)
+        if not (isinstance(data, (list, tuple, np.ndarray)) and len(data)):
```

Comment:
is there a reason for restricting to `list, tuple, np.ndarray` specifically, rather than any sequence? Just curious

Reply:
mostly just an abundance of caution, since I wasn't casting to a known internal format (like `_data = list(data)`) ... but maybe we should?

Reply:
if indexing is happening via your `__getitem__` then casting probably makes sense? definitely not a big deal either way :)

Reply:
I do like the idea of casting to a canonical format, and list seems as good as any. Though note that we eventually might want to do different array sources for 2D and 3D display, and maybe even for different combinations of axes (#3554), so maybe something like a dict of lists etc is what we are moving to? But it's nice that this PR lets us abstract away this question...

Reply:
added the cast